### PR TITLE
Use code generation for fault injection stores

### PR DESCRIPTION
**Pull request by alexshtin**
## What changed?

Use code generation for fault injection stores.

## Why?

They all are the same, and code can be generated. Also removed reflection from hot code path.
Follow up for #5886.

## How did you test it?

Compile.

## Potential risks

No risks.

## Documentation

No.

## Is hotfix candidate?

No.

### CODE REVIEW

File: common/persistence/faultinjection/store_fault_generator.go [hunk 1]
```diff
@@ -56,26 +53,13 @@ func newStoreFaultGenerator(cfg *config.FaultInjectionDataStoreConfig) *storeFau
...
-	pc, _, _, ok := runtime.Caller(1)
-	if !ok {
-		panic("failed to get caller info")
-	}
-	runtimeFunc := runtime.FuncForPC(pc)
-	if runtimeFunc == nil {
-		panic("failed to get runtime function")
-	}
-	parts := strings.Split(runtimeFunc.Name(), ".")
-	methodName := parts[len(parts)-1]
```
Comment (alexshtin) [inline common/persistence/faultinjection/store_fault_generator.go:75]: No more cool reflection.

File: cmd/tools/faultinjection/main.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+			pName = "ctx"
+		} else if strings.HasSuffix(mT.In(i).String(), "Request") {
+			pName = "request"
+		}
+		inParams += fmt.Sprintf("\t %s %s, \n", pName, mT.In(i).String())
+		inVars += fmt.Sprintf("%s, ", pName)
+	}
+	if len(inParams) > 2 {
+		inVars = inVars[:len(inVars)-2]
+	}
```
Comment (dnr) [inline cmd/tools/faultinjection/main.go:159]: what is this doing? (maybe add a comment?)

File: cmd/tools/faultinjection/main.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	inParams := ""
+	inVars := ""
+	for i := 0; i < mT.NumIn(); i++ {
+		pName := fmt.Sprintf("p%d", i)
+		if mT.In(i).String() == "context.Context" {
+			pName = "ctx"
+		} else if strings.HasSuffix(mT.In(i).String(), "Request") {
+			pName = "request"
+		}
+		inParams += fmt.Sprintf("\t %s %s, \n", pName, mT.In(i).String())
```
Comment (dnr) [inline cmd/tools/faultinjection/main.go:154]: you could probably do the whitespace with some template code. but fine either way

**Review by dnr — COMMENTED** [review]
that's less code than I expected!

File: cmd/tools/faultinjection/main.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	var lines []string
+	for _, line := range strings.Split(string(text), "\n") {
+		lines = append(lines, strings.TrimRight("// "+line, " "))
+	}
+	return strings.Join(lines, "\n") + "\n"
+}
+func panicIfErr(err error) {
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply cmd/tools/faultinjection/main.go]: I am all for it but you need to update `rpcwrappers` too.

File: cmd/tools/genfaultinjection/main.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		outParams = fmt.Sprintf("(%s)", outParams)
+	}
+
+	fields := map[string]string{
+		"StoreName":    store.Name(),
+		"MethodName":   m.Name,
+		"InParams":     inParams,
+		"InVars":       inVars,
+		"OutParams":    outParams,
+		"NumOutParams": fmt.Sprintf("%d", mT.NumOut()-1),
```
Comment (dnr) [inline cmd/tools/genfaultinjection/main.go:181]: I forgot this one: you can just use an integer here (with type any). everything is automatically stringified

File: cmd/tools/faultinjection/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dnr) [reply cmd/tools/faultinjection/main.go]: will do when we make a common library

**Review by dnr — APPROVED** [review]
one more question: is the generated code gofmt-clean? it's not essential but it would be nice if it's easy to do with a few tweaks

**Comment by alexshtin** [comment]
Took me extra 15 min to make it gofmt-ed.

**Pull request merged**